File: src/backend/app/utils/timing.py
"""Timing instrumentation for latency analysis.

Usage:
    from app.utils.timing import log_timing, request_id_var

    # At request start
    request_id_var.set(str(uuid.uuid4())[:8])
    log_timing("T1", "request_received")

    # During processing
    log_timing("T2", "llm_start", iteration=1)
    log_timing("T3", "llm_complete", iteration=1, tokens=150)

Output format:
    [TIMING] request_id=abc123 point=T1 ts=1234567890.123 label="request_received"
"""

import logging
import time
from contextvars import ContextVar

# Context variable for request tracking across async calls
request_id_var: ContextVar[str] = ContextVar("request_id", default="unknown")

# Dedicated logger for timing - easy to filter in logs
logger = logging.getLogger("timing")


def log_timing(point: str, label: str, **extra) -> float:
    """Log a timing point with consistent format.

    Args:
        point: Timing point identifier (T1, T2, S1, etc.)
        label: Human-readable label for the timing point
        **extra: Additional key-value pairs to include in log

    Returns:
        The timestamp that was logged (for calculating deltas)
    """
    ts = time.time()
    extra_str = " ".join(f"{k}={v}" for k, v in extra.items())
    logger.info(
        f"[TIMING] request_id={request_id_var.get()} "
        f"point={point} ts={ts:.3f} label={label!r}"
        + (f" {extra_str}" if extra_str else "")
    )
    return ts


def log_timing_delta(point: str, label: str, start_ts: float, **extra) -> float:
    """Log a timing point with delta from a start timestamp.

    Args:
        point: Timing point identifier
        label: Human-readable label
        start_ts: Starting timestamp to calculate delta from
        **extra: Additional key-value pairs

    Returns:
        The current timestamp
    """
    ts = time.time()
    delta_ms = (ts - start_ts) * 1000
    extra_str = " ".join(f"{k}={v}" for k, v in extra.items())
    logger.info(
        f"[TIMING] request_id={request_id_var.get()} "
        f"point={point} ts={ts:.3f} label={label!r} delta_ms={delta_ms:.1f}"
        + (f" {extra_str}" if extra_str else "")
    )
    return ts


class TimingContext:
    """Context manager for timing a block of code.

    Usage:
        with TimingContext("T4", "T5", "tts_call", text_len=45) as ctx:
            result = await some_operation()
        # Automatically logs T4 at start, T5 at end with delta
    """

    def __init__(self, start_point: str, end_point: str, label: str, **extra):
        self.start_point = start_point
        self.end_point = end_point
        self.label = label
        self.extra = extra
        self.start_ts: float = 0
        self.end_ts: float = 0

    def __enter__(self):
        self.start_ts = log_timing(self.start_point, f"{self.label}_start", **self.extra)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.end_ts = log_timing_delta(
            self.end_point,
            f"{self.label}_complete",
            self.start_ts,
            **self.extra
        )
        return False

    async def __aenter__(self):
        return self.__enter__()

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        return self.__exit__(exc_type, exc_val, exc_tb)

    @property
    def delta_ms(self) -> float:
        """Get the elapsed time in milliseconds."""
        return (self.end_ts - self.start_ts) * 1000

File: src/frontend/src/utils/timing.ts
/**
 * Timing instrumentation for latency analysis.
 *
 * Usage:
 *   import { logTiming, generateRequestId, timingState } from '@/utils/timing'
 *
 *   // At request start
 *   const requestId = generateRequestId()
 *   timingState.setRequestId(requestId)
 *   logTiming('T0', 'request_sent')
 *
 *   // During processing
 *   logTiming('T10', 'response_received')
 *   logTiming('T11', 'audio_playback_start')
 *
 * Events are logged to console AND posted to /api/timing/log for aggregation.
 */

const TIMING_API_BASE = '/api/timing'

// Current request context
class TimingState {
  private requestId: string = 'unknown'
  private t0Timestamp: number = 0
  private s0Timestamp: number = 0

  setRequestId(id: string) {
    this.requestId = id
  }

  getRequestId(): string {
    return this.requestId
  }

  setT0(ts: number) {
    this.t0Timestamp = ts
  }

  getT0(): number {
    return this.t0Timestamp
  }

  setS0(ts: number) {
    this.s0Timestamp = ts
  }

  getS0(): number {
    return this.s0Timestamp
  }

  reset() {
    this.requestId = 'unknown'
    this.t0Timestamp = 0
    this.s0Timestamp = 0
  }
}

export const timingState = new TimingState()

/**
 * Generate a short unique request ID for timing correlation.
 */
export function generateRequestId(): string {
  return crypto.randomUUID().slice(0, 8)
}

/**
 * Post timing event to backend (fire-and-forget).
 */
function postTimingEvent(
  requestId: string,
  point: string,
  ts: number,
  label: string,
  extra?: Record<string, string | number | boolean>
): void {
  // Fire-and-forget POST - don't await, don't block
  fetch(`${TIMING_API_BASE}/log`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      request_id: requestId,
      point,
      ts,
      label,
      extra: extra || null,
    }),
  }).catch(() => {
    // Silently ignore failures - timing shouldn't break the app
  })
}

/**
 * Log a timing point with consistent format.
 *
 * @param point - Timing point identifier (T0, T10, T11, S0, S1, S5)
 * @param label - Human-readable label
 * @param extra - Additional key-value pairs
 * @returns The timestamp that was logged
 */
export function logTiming(
  point: string,
  label: string,
  extra: Record<string, string | number | boolean> = {}
): number {
  const ts = Date.now() / 1000 // Match backend format (Unix timestamp)
  const requestId = timingState.getRequestId()

  // Console log for local debugging
  const extraStr = Object.entries(extra)
    .map(([k, v]) => `${k}=${v}`)
    .join(' ')

  console.log(
    `[TIMING] request_id=${requestId} point=${point} ts=${ts.toFixed(3)} label="${label}"` +
      (extraStr ? ` ${extraStr}` : '')
  )

  // Post to backend for aggregation
  postTimingEvent(requestId, point, ts, label, Object.keys(extra).length > 0 ? extra : undefined)

  return ts
}

/**
 * Log a timing point with delta from a start timestamp.
 */
export function logTimingDelta(
  point: string,
  label: string,
  startTs: number,
  extra: Record<string, string | number | boolean> = {}
): number {
  const ts = Date.now() / 1000
  const deltaMs = (ts - startTs) * 1000

  return logTiming(point, label, { ...extra, delta_ms: deltaMs.toFixed(1) })
}

/**
 * Calculate and log total latency from T0.
 */
export function logTotalLatency(point: string, label: string): number {
  const t0 = timingState.getT0()
  if (t0 === 0) {
    return logTiming(point, label)
  }
  return logTimingDelta(point, label, t0, { total: true })
}
